reicast/apple/emulator-osx/emulator-osx/AppDelegate.mm

```diff
@@ -0,0 +1,272 @@
+    // Must multiply by the scale factor to account for retina displays
+    displayPixelSize.width *= screen.backingScaleFactor;
+    displayPixelSize.height *= screen.backingScaleFactor;
+    CGSize displayPhysicalSize = CGDisplayScreenSize([[description objectForKey:@"NSScreenNumber"] unsignedIntValue]);
+    screen_dpi = (int)(displayPixelSize.width / displayPhysicalSize.width) * 25.4f;
+    NSLog(@"displayPixelSize: %@  displayPhysicalSize: %@  screen_dpi: %d", NSStringFromSize(displayPixelSize), NSStringFromSize(displayPhysicalSize), screen_dpi);
+}
+
+static BOOL _isShuttingDownEmulator = NO;
+- (void)shutdownEmulator {
+    if (_isShuttingDownEmulator) {
+        return;
+    }
+    _isShuttingDownEmulator = YES;
+    
+#if FEAT_HAS_SERIAL_TTY 
+    if (_removePTYSymlink) {
+        unlink(settings.debug.VirtualSerialPortFile.c_str());
+    }
+#endif
+    
+    // Shut down emulator
+    if (virtualDreamcast && sh4_cpu->IsRunning()) {
+        virtualDreamcast->Stop([] {
+            g_GUIRenderer->Stop();
+            reicast_term();
+            _isShuttingDownEmulator = NO;
+        });
+    } else {
+        g_GUIRenderer->Stop();
+        reicast_term();
+        _isShuttingDownEmulator = NO;
+    }
+}
+
+- (void)alertAndTerminateWithMessage:(NSString *)message {
+    NSAlert *alert = [[NSAlert alloc] init];
+    alert.alertStyle = NSCriticalAlertStyle;
+    alert.messageText = message;
+    [alert addButtonWithTitle:@"Exit"];
+    [alert runModal];
+    [NSApp terminate:nil];
+}
+
+#pragma mark - Reicast OS Functions -
+
+void os_LaunchFromURL(const string& url) {
+    NSString *urlString = [NSString stringWithUTF8String:url.c_str()];
+    [[NSWorkspace sharedWorkspace] openURL:[NSURL URLWithString:urlString]];
+}
+
+void os_SetWindowText(const char * text) {
+    puts(text);
+}
+
+// Code that runs at the beginning of every frame
+void os_DoEvents() {
+    // Left empty on purpose as we have nothing to run here
+}
+
+// Handle controller input
+void UpdateInputState(u32 port) {
+    // TODO: BEN not sure what, if anything, we need here
+}
+
+// Called by libswirl to create the emulator window
+void os_CreateWindow() {
+    // Left empty on purpose as we have nothing to run here
+}
+
+// Called by libswirl when OpenGL is initialized
+bool os_gl_init(void* hwnd, void* hdc) {
+    [_sharedInstance.glView.openGLContext makeCurrentContext];
+    gl_resize();
+    return true;
+}
+
+// Called on every frame to swap the buffer
+bool os_gl_swap() {
+    NSOpenGLContext *glContext = _sharedInstance.glView.openGLContext;
+    [glContext makeCurrentContext];
+    [glContext flushBuffer];
+    
+    return true;
+}
+
+// Called when the application terminates
+void os_gl_term() {
+    [_sharedInstance shutdownEmulator];
+}
+
+static void gl_resize() {
+    // Handle HDPI by multiplying the scale factor
+    rend_resize(_glViewSize.width * _backingScaleFactor, _glViewSize.height * _backingScaleFactor);
```

Comment:
is this in pixels? I don't know how scaling works for mac

Reply:
Yep this is in pixels. Normally everything is done with "points", which is part of the problem here. When initializing the Mac's GL view with a size of 640x480, its framebuffer gets set to 640x480 in points, which is actually 1280x960 in pixels when using a "retina" display aka a display with backing scale factor of 2.0. However the Reicast renderer is thinking only in pixels, so without this rend_resize call multiplying by the scale factor, the renderer output takes up only the bottom left quarter of the view on retina displays.

This just corrects for that and ensures that the framebuffer of the gl view and of the renderer are the same on high dpi displays.